### PR TITLE
✨ Implement "clusterctl completion" command

### LINKED ISSUE #3595: bash/zsh auto completion for clusterctl
**User Story**

As a developer/user/operator, I would like to use bash/zsh auto completion for clusterctl for improving work efficiency.

**Detailed Description**

I'd like to use bash/zsh auto completion for clusterctl command, like `kubectl completion` command.

```
$ source <(clusterctl completion bash)
```

I believe this will improve the experience for everyone - developers/users/operators.

**Anything else you would like to add:**

None

/kind feature

### PR DESCRIPTION
**Pull request by superbrothers**
**What this PR does / why we need it**:

This PR implements the `clusterctl completion` command that outputs bash/zsh completion code. The `clusterctl completion command` implemented by this PR provides auto completion for subcommands and flags. The next PR will provide dynamic auto completion for arguments of `--kubeconfig-context`, etc.

You can use auto completion for the clusterctl command as follows:
```
# Install bash completion on a Mac using homebrew
brew install bash-completion
printf "\n# Bash completion support\nsource $(brew --prefix)/etc/bash_completion\n" >> $HOME/.bash_profile
source $HOME/.bash_profile

# Load the clusterctl completion code for bash into the current shell
source <(clusterctl completion bash)

# Load the clusterctl completion code for zsh[1] into the current shell
source <(clusterctl completion zsh)
```

This code is based on the `kubectl completion` and` kubeadm completion` commands.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3595

### CODE REVIEW

File: cmd/clusterctl/cmd/completion.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+		# Write bash completion code to a file and source it from .bash_profile
+		clusterctl completion bash > ~/.kube/clusterctl_completion.bash.inc
+		printf "\n# clusterctl shell completion\nsource '$HOME/.kube/clusterctl_completion.bash.inc'\n" >> $HOME/.bash_profile
+		source $HOME/.bash_profile
+
+		# Load the clusterctl completion code for zsh[1] into the current shell
+		source <(clusterctl completion zsh)`)
+
+	completionCmd = &cobra.Command{
+		Use:       "completion SHELL",
```
Comment (benmoss) [inline cmd/clusterctl/cmd/completion.go:81]:
```suggestion
		Use:       "completion [bash|zsh]",
```

File: cmd/clusterctl/cmd/completion.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+func runCompletion(cmd *cobra.Command, args []string) error {
+	if len(args) == 0 {
+		return errors.New("Shell not specified.")
+	}
+
+	if len(args) > 1 {
+		return errors.New("Too many arguments. Expected only the shell type.")
+	}
```
Comment (benmoss) [inline cmd/clusterctl/cmd/completion.go:118]: These are unreachable, cobra validates this all for you

```
./bin/clusterctl completion
Error: accepts 1 arg(s), received 0
./bin/clusterctl completion bash foo
Error: accepts 1 arg(s), received 2
```

File: cmd/clusterctl/cmd/completion.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	if !found {
+		return fmt.Errorf("Unsupported shell type %q.", args[0])
+	}
+
+	return run(cmd.Parent())
+}
+
+func runCompletionBash(cmd *cobra.Command) error {
+	cmd.GenBashCompletion(os.Stdout)
+	return nil
```
Comment (benmoss) [inline cmd/clusterctl/cmd/completion.go:130]:
```suggestion
	return cmd.GenBashCompletion(os.Stdout)
```

File: cmd/clusterctl/cmd/completion.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+func runCompletionZsh(cmd *cobra.Command) error {
+	zshHead := "#compdef clusterctl\n"
```
Comment (benmoss) [inline cmd/clusterctl/cmd/completion.go:134]: Nit but I'd make these all static strings into constants outside the method body

File: cmd/clusterctl/cmd/completion.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+		Short:     "Output shell completion code for the specified shell (bash or zsh)",
+		Long:      LongDesc(completionLong),
+		Example:   completionExample,
+		Args:      cobra.ExactArgs(1),
+		RunE:      runCompletion,
+		ValidArgs: GetSupportedShells(),
+	}
+)
+
+var (
```
Comment (benmoss) [inline cmd/clusterctl/cmd/completion.go:91]: No need for another var section
```suggestion

```

File: cmd/clusterctl/cmd/completion.go [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+	-e "s/${LWORD}__ltrim_colon_completions${RWORD}/__clusterctl_ltrim_colon_completions/g" \
+	-e "s/${LWORD}compgen${RWORD}/__clusterctl_compgen/g" \
+	-e "s/${LWORD}compopt${RWORD}/__clusterctl_compopt/g" \
+	-e "s/${LWORD}declare${RWORD}/builtin declare/g" \
+	-e "s/\\\$(type${RWORD}/\$(__clusterctl_type/g" \
+	<<'BASH_COMPLETION_EOF'
+`
+	fmt.Print(zshInitialization)
+
+	cmd.GenBashCompletion(os.Stdout)
```
Comment (benmoss) [inline cmd/clusterctl/cmd/completion.go:230]:
```suggestion
	if err := cmd.GenBashCompletion(os.Stdout); err != nil {
		return err
	}
```

**Review by benmoss — COMMENTED** [review]
Some more comments!

File: cmd/clusterctl/cmd/completion.go (hunk 6 above)
Comment (superbrothers) [reply cmd/clusterctl/cmd/completion.go]: Fixed.

File: cmd/clusterctl/cmd/completion.go (hunk 2 above)
Comment (superbrothers) [reply cmd/clusterctl/cmd/completion.go]: Fixed.

File: cmd/clusterctl/cmd/completion.go (hunk 4 above)
Comment (superbrothers) [reply cmd/clusterctl/cmd/completion.go]: Fixed by https://github.com/kubernetes-sigs/cluster-api/pull/3622/commits/45842cee955acfbfdd7998b1e2240bdca09d3a00.

File: cmd/clusterctl/cmd/completion.go (hunk 5 above)
Comment (superbrothers) [reply cmd/clusterctl/cmd/completion.go]: Fixed by https://github.com/kubernetes-sigs/cluster-api/pull/3622/commits/9ad122d707ed336e2c86d720f305d25140370e53.

File: cmd/clusterctl/cmd/completion.go (hunk 3 above)
Comment (superbrothers) [reply cmd/clusterctl/cmd/completion.go]: Fixed by https://github.com/kubernetes-sigs/cluster-api/pull/3622/commits/caa9d0c9d43de72d4a9433bf67c35a8707006eec.

**Comment by wfernandes** [comment]
/assign
reviewing

**Review by wfernandes — COMMENTED** [review]
Thanks for this PR @superbrothers! I wanted to add this in myself (clearly you beat me to it 😉 ).

We will need some documentation for this command. I suggest adding it here https://github.com/kubernetes-sigs/cluster-api/tree/2e1d3770c320869fd96f9c82337dd3a328f8ff1a/docs/book/src/clusterctl/commands which will show up in our book https://cluster-api.sigs.k8s.io/clusterctl/commands/commands.html.

Also it seems that the bash completion is generated using cobra and the zsh completion is generated from bash. So I'm guessing if new commands are added to clusterctl, do users need to do rerun the completion command? Perhaps worth documenting.

File: cmd/clusterctl/cmd/completion.go [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+		clusterctl commands. This can be done by sourcing it from the
+		.bash_profile.
+
+		Note: this requires the bash-completion framework.
+
+		To install it on Mac use homebrew:
+		    $ brew install bash-completion
+		Once installed, bash_completion must be evaluated. This can be done by
+		adding the following line to the .bash_profile
+		    $ source $(brew --prefix)/etc/bash_completion
```
[the remark this answers is not in the document]
Comment (superbrothers) [reply cmd/clusterctl/cmd/completion.go]: Fix by a5d985d754c5aab44ab654c32f0de435c044a0e0.

**Comment by superbrothers** [comment]
@wfernandes I've added documentation for the `clusterctl completion` command at https://github.com/kubernetes-sigs/cluster-api/pull/3622/commits/e54af12cedb2f9d446dbd28f51418237b7efafd0.

**Comment by superbrothers** [comment]
> So I'm guessing if new commands are added to clusterctl, do users need to do rerun the completion command? Perhaps worth documenting.

@wfernandes We source the autocompletion code generated each time we log in by doing the following, so you don't need to do anything when new commands are added to clusterctl.
```
# bash
source <(clusterctl completion bash)

# zsh
source <(clusterctl completion zsh)
```

**Comment by wfernandes** [comment]
@superbrothers Thanks for the updates and the explanation. And thanks for this PR!
/approve

**Comment by wfernandes** [comment]
I suggest updating the description to `Fixes #3595` so that the issue closes when this PR gets merged.

File: cmd/clusterctl/cmd/completion.go [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+const completionBoilerPlate = `
+# Copyright 2020 The Kubernetes Authors.
```
Comment (vincepri) [inline cmd/clusterctl/cmd/completion.go:27]: Nit: For generated files, we can remove the year, see https://github.com/kubernetes-sigs/cluster-api/blob/master/hack/boilerplate/boilerplate.generatego.txt

**Review by vincepri — APPROVED** [review]
/approve

**Comment by vincepri** [comment]
Really great job, thanks for doing this!

/milestone v0.3.10

**Pull request merged**